magicgui/widgets/_table.py

```diff
@@ -0,0 +1,134 @@
+from typing import Any, Tuple
+
+from ._bases import ValueWidget
+from ._concrete import backend_widget
+
+
+def _is_dataframe(obj) -> bool:
+    try:
+        import pandas
+
+        pandas.DataFrame.from_dict
+        return isinstance(obj, pandas.DataFrame)
+    except ImportError:
+        return False
+
+
+def _is_numpy_array(obj) -> bool:
+    try:
+        import numpy
+
+        return isinstance(obj, numpy.ndarray)
+    except ImportError:
+        return False
+
+
+def _from_nested_index_dict(data) -> dict:
+    from collections import defaultdict
+
+    new_data: dict = defaultdict(dict)
+    for index, s in data.items():
+        for col, v in s.items():
+            new_data[col][index] = v
+    return new_data
+
+
+def _from_nested_column_dict(data):
+    _index = {frozenset(i) for i in data.values()}
+    if len(_index) > 1:
+        raise ValueError(
+            "All row-dicts must have the same keys. "
+            "Install pandas for better table-from-dict support."
+        )
+    index = tuple(sorted(list(_index)[0]))
+
+    new_data = []
+    for col, s in data.items():
+        new_data.append([s[i] for i in index])
+    return list(zip(*new_data)), index
+
+
+def _from_dict(
+    data, orient="columns", dtype=None, columns=None
+) -> Tuple[list, tuple, tuple]:
+    """Construct _MguiTable from dict of array-like or dicts.
+
+    logic from pandas.DataFrame.from_dict
+    """
+    try:
+        import pandas
+
+        df = pandas.DataFrame.from_dict(
+            data, orient=orient, dtype=dtype, columns=columns
+        )
+        return df.values, tuple(df.index), tuple(df.columns)
+    except ImportError:
+        pass
+
+    index: tuple = ()
+    orient = orient.lower()
+    if orient == "index":
+        if len(data) > 0:
+            # TODO: this is inaccessible, since Table.value still assumes orient=columns
+            # but it may not work
+            if isinstance(list(data.values())[0], dict):
+                data = _from_nested_index_dict(data)
+            else:
+                data, index = list(data.values()), tuple(data.keys())
+    elif orient == "columns":
+        if columns is not None:
+            raise ValueError("cannot use columns parameter with orient='columns'")
+
+        columns = tuple(data.keys())
+        if isinstance(list(data.values())[0], dict):
+            data, index = _from_nested_column_dict(data)
+        else:
+            data = list(zip(*data.values()))
+
+    else:  # pragma: no cover
+        raise ValueError("only recognize index or columns for orient")
+    return data, index, columns or ()
+
+
+def _value_to_table_data(value: Any) -> Tuple[list, tuple, tuple]:
+    """Convert arbitrary value into appropriate table data format."""
```

Comment:
imho this docstring should be expanded. `value` is hardly "arbitrary", but rather one of several formats, which I don't see documented anywhere. Is dict-of-dicts common? I have never seen it before and I don't love it. And I haven't seen the 3-tuple of (data, index, columns), either... Do we not want to use the dict-of-columns as the default format, which would then make it directly compatible with napari's properties tables? (Not to mention pandas?)

Reply:
yeah, def... will expand the docstring.
All these formats are things that I found when digging through the formats that a DataFrame takes.  The dict-of-dicts is what you get if you do [`DataFrame.to_dict()`](https://github.com/pandas-dev/pandas/blob/aad85ad1e39568f87cdae38fece41cd2370234a7/pandas/core/frame.py#L1496-L1497)  (i looked through their API for the most "pure python type" version of a dataframe, and that's what I found...  it does seem like the only way to get data, headers, and index headers all in one, so.. maybe people use it?)
As for the default format... (a 2d data array, a list of headers and a list of row indexes), that was borrowed from the [internal Dataframe representation](https://github.com/pandas-dev/pandas/blob/aad85ad1e39568f87cdae38fece41cd2370234a7/pandas/core/frame.py#L506-L508), which has a 2D array of `.values`, a 1D `.index` and a 1D `.columns`.  so compatibility with pandas was pretty much my first priority here.  That said, I'm happy to have some different representation be the default when you do `table.value`!  thoughts?

Reply:
I should add, I'm more than happy with a dict of columns... but that excludes row headers (should someone want them)

Reply:
(also, this can all wait 😄 ... will do 0.2.0 without and we can discuss later)

Reply:
> internal Dataframe representation

Well, I think that's glossing over some rather major subtleties because of dtypes — see this rather excellent blog post from Uwe Korn:

https://uwekorn.com/2020/05/24/the-one-pandas-internal.html

imho row headers are less useful and accessible via `reset_index` anyway...

Reply:
sorry, I wasn't suggesting that pandas _actually_ maintains something like a numpy array and two lists internally... but it does take it as an input, and if you do `df.values`, `df.index`, `df.columns`, that's what you get.
as mentioned, I'm happy to use a dict of column arrays.  but even if row headers are less useful, there should be a way for someone to provide them and query them here (outside of using pandas), so how do you want that to look?

Reply:
@jni ... don't want this to get too far off the radar.  It's mostly waiting on your approval of the API at this point.  I don't care too much what the data structure is but I wanted it to be:
- pure python
- easy to convert to/from a dataframe
- able to represent table data, row headers, and column headers

beyond that, whatever you want is fine with me

Reply:
@tlambert03 

> even if row headers are less useful, there should be a way for someone to provide them and query them here (outside of using pandas),

I'd aim to mirror the `read_csv` API, with the index included as a column and `index_col='column_name'` as part of the API. How does that sound?

Reply:
> I'd aim to mirror the read_csv API, with the index included as a column and index_col='column_name' as part of the API. How does that sound?

I _think_ I'm ok with that, but i want to just make sure we're clear on what part of the API we're discussing here. What this really comes down to (at the moment) is not what the user has to provide to magicgui (They're going to be able to set the value with a lot of standard formats... when setting `widget.value` ) it's what object the user gets _back_ when retrieving the value from the widget.  So are you proposing this?

```python
# one of many valid input formats
widget = Table()
widget.value = [[1,2,3],[4,5,6]]
# get the value back
widget.value

# returns
{0: [1,4], 1: [2,5], 2: [3,6], 'index': [0,1], 'index_col': 'index'}
```

something tells me that's _not_ what you're imagining (I'm not a fan myself).  I'm quite happy to return a tuple as I started with `widget.value -> ([[1,2,3],[4,5,6]], [], [])`, or a `NamedTuple`...  I really don't want to require someone to depend on pandas just to provide a simple 2x2 table with text.  I feel rather firm on that for this stage of development.  I made the `TableData` object to have something like a named tuple that would provide `data.value`, `data.index`, `data.columns` so that if we _do_ go to a dataframe in the future, the user just gets _more_ API, (and the existing API doesn't break).

If you don't like any of the dict return shown above, the tuple return value, or the `TableData` object, please provide an example of the pure python object that you want the user to get back when they query the current value of the widget.

Reply:
It occurs to me that you might look at the example above and say "if they give me `widget.value = [[1,2,3],[4,5,6]]` then that's what they should get back".  So please also imagine that they provided a more complex object in, like one of the supported dict formats that also provides header info

Reply:
another option is that we extend the `ValueWidget` API just for `Table` widgets, so that they can get and set `widget.index` and `widget.columns`, reducing `widget.value` to a simple 2x2 array that never has header information

Reply:
⬆️ I think this is the best approach... and will work on that unless you object

Reply:
On phone, but briefly, imho it would be great to have .values (or .data, or .properties, I'm agnostic here and I do think it might be good to differentiate from a common API if we don't conform to it) be a dict-of-columns, and *optionally* a string attribute index_col that lets the user know which is the "special" column, if any, that should be used as index. 

Does that meet your requirements for everything that needs to be represented?